File: 02_stack/05_car_fleet/car_fleet.cpp
#include <vector>
#include <algorithm>

using std::vector;

class Solution
{
  public:
    struct Car
    {
        int   position_;
        float time_;

        Car(int position, float time) : position_(position), time_(time)
        {
        }
    };

    int carFleet(int target, vector<int>& position, vector<int>& speed)
    {
        vector<Car> cars;
        for (size_t i = 0; i < position.size(); i++)
        {
            cars.push_back(
                Car(position[i],
                    (target - position[i]) / static_cast<float>(speed[i])));
        }
        std::sort(cars.begin(), cars.end(), [](const Car& a, const Car& b) {
            return a.position_ < b.position_;
        });

        int   fleet_amount = 1;
        float last_time    = cars.back().time_;

        for (auto car = ++cars.rbegin(); car != cars.rend(); ++car)
        {
            if (car->time_ > last_time)
            {
                ++fleet_amount;
                last_time = car->time_;
            }
        }
        return fleet_amount;
    }
};

#include <iostream>

int main()
{
    Solution    solution;
    int         target   = 12;
    vector<int> position = {10, 8, 0, 5, 3};
    vector<int> speed    = {2, 4, 1, 1, 3};
    std::cout << solution.carFleet(target, position, speed) << "\n";
}

File: NeetCode/02_stack/05_car_fleet/car_fleet.py
from typing import List


class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        cars = [(position[i], (target - position[i]) / speed[i])
                for i in range(len(position))]
        cars.sort(key=lambda car: car[0])
        # print(cars)

        fleet_amount = 1
        last_time = cars.pop()[1]
        for position, time in reversed(cars):
            if time > last_time:  # Doesn't catch up the last fleet
                fleet_amount += 1
                last_time = time
        return fleet_amount


if __name__ == "__main__":
    solution = Solution()
    position = [10, 8, 0, 5, 3]
    speed = [2, 4, 1, 1, 3]
    target = 12
    print(solution.carFleet(target, position, speed))
